x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/action/ShardFollowTaskReplicationTests.java

```diff
@@ -0,0 +1,225 @@
+        for (Map.Entry<String, String> typeMapping : mappings.entrySet()) {
+            metaData.putMapping(typeMapping.getKey(), typeMapping.getValue());
+        }
+        return metaData.build();
+    }
+
+    private ReplicationGroup createFollowGroup(int replicas) throws IOException {
+        Settings.Builder settingsBuilder = Settings.builder();
+        settingsBuilder.put(CcrSettings.CCR_FOLLOWING_INDEX_SETTING.getKey(), true);
+        return createGroup(replicas, settingsBuilder.build());
+    }
+
+    private ShardFollowNodeTask createShardFollowTask(ReplicationGroup leaderGroup, ReplicationGroup followerGroup) {
+        ShardFollowTask params = new ShardFollowTask(null, new ShardId("follow_index", "", 0),
+            new ShardId("leader_index", "", 0), 1024, 1, Long.MAX_VALUE, 1, 10240,
+            TimeValue.timeValueMillis(10), TimeValue.timeValueMillis(10), Collections.emptyMap());
+
+        BiConsumer<TimeValue, Runnable> scheduler = (delay, task) -> {
+            try {
+                Thread.sleep(delay.millis());
+            } catch (InterruptedException e) {
+                throw new AssertionError(e);
+            }
+            task.run();
+        };
+        AtomicBoolean stopped = new AtomicBoolean(false);
+        return new ShardFollowNodeTask(1L, "type", ShardFollowTask.NAME, "description", null, Collections.emptyMap(), params, scheduler) {
+            @Override
+            protected void updateMapping(LongConsumer handler) {
+                // noop, as mapping updates are not tested
+                handler.accept(1L);
+            }
+
+            @Override
+            protected void innerSendBulkShardOperationsRequest(List<Translog.Operation> operations, LongConsumer handler,
+                                                               Consumer<Exception> errorHandler) {
+                Runnable task = () -> {
+                    try {
+                        IndexShard primaryShard = followerGroup.getPrimary();
+                        TransportWriteAction.WritePrimaryResult<BulkShardOperationsRequest, BulkShardOperationsResponse> result =
+                            TransportBulkShardOperationsAction.shardOperationOnPrimary(primaryShard.shardId(), operations,
+                                primaryShard, logger);
+                        List<Translog.Operation> replicaOps = result.replicaRequest().getOperations();
+                        for (IndexShard replicaShard : followerGroup.getReplicas()) {
+                            TransportBulkShardOperationsAction.applyTranslogOperations(replicaOps, replicaShard, Origin.REPLICA);
+                        }
+                        handler.accept(primaryShard.getGlobalCheckpoint());
+                    } catch (Exception e) {
+                        errorHandler.accept(e);
+                    }
+                };
+                Thread thread = new Thread(task);
```

Comment:
Can we make sure that this thread is completed when asserting stuffs in the test?

Reply:
I'm not sure how we can do that here. If I add a `thread#join()` after the start has started here then it results into a deadlock, because the thread can call this method method recursively (indirectly, via handleReadResponse() -> coordinateReads()).

Reply:
Sorry, I should have been more clear. I feel a bit uncomfortable when we do not care about the termination of this thread. Is it ok if we submit this task to the generic threadPool instead?

Reply:
That makes sense. I pushed this: 90318b24e2d516bb85484664f5c6d3a67a656260